README.md
[![codecov](https://codecov.io/gh/molgenis/molgenis-ui-form/branch/master/graph/badge.svg)](https://codecov.io/gh/molgenis/molgenis-ui-form)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

# molgenis-ui-filter

## Filter Container

The filter container takes a descriptive input array to create a set of filters. It will also require a 2 way bound output object to receive the filters state

```
<filter-container
  v-model="selections"
  :filters="filters"
  :filters-shown="filtersShown"
  :can-edit="true"
/>
```

### Properties

- `selections`: Will hold the settings set in filters.
- `filters`: A definition list of all possible filters.
- `filtersShown`: A array of filter names that will be shown.
- `can-edit`: if true you will be able to add, move and remove filters.

### Events

- `update`: this event is fired when filtersShown is updated, it will return a new list of selected filters

### Filter input object specifications

```
filters: [ {
    name: 'search',
    label: 'Search',
    description: 'search by name',
    initiallyCollapsed: true,
    placeholder: 'test',
    type: 'string-filter'
}, {
    name: 'country',
    label: 'Countries',
    initiallyCollapsed: false,
    options: [{ value: 'nl', text: 'Nederland' }, { value: 'de', text: 'Duitsland' }, { value: 'be', text: 'België' }],
    type: 'checkbox-filter'
}]
```

Every filter will have the following global properties:

| name                 | required | default value | description | 
|----------------------|----------|---------------|-------------|
| name                 | true     | *na*          | string: name and id of the filter |
| label                | false    |               | string: name of the filter as used by the UI |
| description          | false    |               | descriptive text that is used by the UI |
| collapsed            | false    | true          | boolean: Open/Closed state of the card that holds the filter |
| collapsable          | false    | true          | boolean: Can the card be collapsed. This will force the card to be open by default if set to false |
| type                 | true     | *na*          | vue object that renders the filter |
| value                | false    | empty         | the (pre)selected option(s) | 

The `type` field takes a VUE object that renders a filter. Note that every object can have additional filter properties

### Additional filter properties
#### CheckboxFilter 

| name                 | required | default value | description | 
|----------------------|----------|---------------|-------------|
| options              | true     | *na*          | describe the checkboxes. This is a function holding a promise holding the options (see example below) |
| bulkOperation        | false    | true          | adds 'select all' and or 'Deselect all' button |
| maxVisibleOptions    | false    | undefined *(show all options)* | limit the amount of initially shown options | 

*options example*
```
() => {
    return new Promise(
        function (resolve) {
            resolve([{ value: 'value1', text: 'Checkbox 1' }, { value: 'value2', text: 'Checkbox 2' }])
        }
    )
}
```

#### DateTimefilter

| name                 | required | default value    | description | 
|----------------------|----------|------------------|-------------|
| max                  | false    | null `: string`  | Set the minimal date you can select, needs to be able to be parsed by the Date object |
| min                  | false    | null `: string`  | Set the maximum date you can select, needs to be able to be parsed by the Date object | 
| opens                | false    | 'right' `: string` | Set the alignment of the popup. Possible options: 'left', 'right', 'center' | 
| time                 | false    | true `: boolean` | Add a time selection to the data time picker | 
| range                | false    | true `: boolean` | Add a range selection to be able to make a from-to selection | 

#### Multifilter

The multifilter is a composite filter. It uses a searchbox, which searches for options. 
The usecase is to generate a list of checkboxes based on a user query (capped to first 100) when the options are of _n_ length.

| name                 | required | default value | description | 
|----------------------|----------|---------------|-------------|
| options              | true     | *na* `: Function` | describe the checkboxes. This is a function returns a promise which can query and return options (see example below) |
| placeholder          | false    | empty `: String` | fill the placeholder in the searchbox | 
| maxVisibleOptions    | false    | 10 `: Number` | limit the amount of initially shown options by _n_ | 
| initialDisplayItems  | false    | 5 `: Number` | Initial set of items to be displayed | 

*options example*
```
(params: object) => {
  const data = await axios.get(url, { params })

  return Promise.resolve(
    data.data.items.map((i: any) => {
      return { value: i.data[*idName*], text: i.data[*attrName*] }
    })
  )
}
```

#### StringFilter
| name                 | required | default value | description | 
|----------------------|----------|---------------|-------------|
| placeholder          | false    |               | Placeholder in the input field |

#### NumberFilter, RangeFilter
| name                 | required | default value | description | 
|----------------------|----------|---------------|-------------|
| min                  | false    |               | minimal posible value |
| max                  | false    |               | maximal posible value |
| step                 | false    | 1             | step size of slider and up/down buttons |
| useSlider            | false    | false         | show a slider to help pick a number |

### Resulting output object

The resulting filter status is an object that holds the name fields of the filter description object and the resulting selected values. Note the 2 way vue binding.
Feed in the default values by setting them here on object creation.

```
selections: {
  search: '',
  country: ['nl']
},
```

## Project setup
```
yarn install
```

### Compiles and hot-reloads for development
```
yarn run serve
```

### Compiles and minifies for production
```
yarn run build
```

### Lints and fixes files
```
yarn run lint
```

### Run your end-to-end tests
```
yarn run test:e2e
```

### Run your unit tests
```
yarn run test:unit
```

### How to commit
We use conventional commits to generate changelogs and release notes. Please check: https://www.conventionalcommits.org/

**Example**
```
git commit file.ext -m "fix(file.ext): fixes something"
```

### How to publish
Each time a PR is merged a release will be done to NPM. The CHANGELOG.md and GitHub release will be ammended. 

The version of the package is based upon convential commits. Check: http://commitizen.github.io/cz-cli/. 


CHANGELOG.md
## [1.11.1](https://github.com/molgenis/molgenis-ui-filter/compare/v1.11.0...v1.11.1) (2020-07-02)


### Bug Fixes

* fixed empty string for one input field not resulting in a 'or less, or more' . Added stronger typing. ([5067c04](https://github.com/molgenis/molgenis-ui-filter/commit/5067c04))

# [1.11.0](https://github.com/molgenis/molgenis-ui-filter/compare/v1.10.1...v1.11.0) (2020-07-02)


### Features

* release new version ([eb62b47](https://github.com/molgenis/molgenis-ui-filter/commit/eb62b47))

## [1.10.1](https://github.com/molgenis/molgenis-ui-filter/compare/v1.10.0...v1.10.1) (2020-06-25)


### Bug Fixes

* make the range input react correctly to number that use a dot ([a214ba4](https://github.com/molgenis/molgenis-ui-filter/commit/a214ba4))

# [1.10.0](https://github.com/molgenis/molgenis-ui-filter/compare/v1.9.3...v1.10.0) (2020-06-25)


### Bug Fixes

* **search:** query was not working. ([54c2a38](https://github.com/molgenis/molgenis-ui-filter/commit/54c2a38))


### Features

* add initial items to display to multifilter ([3ff67a8](https://github.com/molgenis/molgenis-ui-filter/commit/3ff67a8))

## [1.9.3](https://github.com/molgenis/molgenis-ui-filter/compare/v1.9.2...v1.9.3) (2020-05-19)


### Bug Fixes

* jenkins publish to npm ([e1ddf36](https://github.com/molgenis/molgenis-ui-filter/commit/e1ddf36))

## [1.9.2](https://github.com/molgenis/molgenis-ui-filter/compare/v1.9.1...v1.9.2) (2020-05-15)


### Bug Fixes

* remove unnecessary updating trigger ([4281572](https://github.com/molgenis/molgenis-ui-filter/commit/4281572))
* updated how multifilter reacts, reverted a fix in filtercontainer ([8f15929](https://github.com/molgenis/molgenis-ui-filter/commit/8f15929))

## [1.9.1](https://github.com/molgenis/molgenis-ui-filter/compare/v1.9.0...v1.9.1) (2020-05-12)


### Bug Fixes

* make sure that de add-filter dialog is always reachable ([731db6f](https://github.com/molgenis/molgenis-ui-filter/commit/731db6f))

# [1.9.0](https://github.com/molgenis/molgenis-ui-filter/compare/v1.8.3...v1.9.0) (2020-05-12)


### Bug Fixes

* bind to viewport ([4bc63aa](https://github.com/molgenis/molgenis-ui-filter/commit/4bc63aa))
* e2e test ([2b43787](https://github.com/molgenis/molgenis-ui-filter/commit/2b43787))
* e2e test ([1ac8d21](https://github.com/molgenis/molgenis-ui-filter/commit/1ac8d21))
* run linter ([f725c80](https://github.com/molgenis/molgenis-ui-filter/commit/f725c80))


### Features

* add filters checkbox selections ([509a808](https://github.com/molgenis/molgenis-ui-filter/commit/509a808))
* add styling and compound indenting ([2f214e2](https://github.com/molgenis/molgenis-ui-filter/commit/2f214e2))
* compound handeling in add filter dialog ([ae436f7](https://github.com/molgenis/molgenis-ui-filter/commit/ae436f7))

## [1.8.3](https://github.com/molgenis/molgenis-ui-filter/compare/v1.8.2...v1.8.3) (2020-04-30)


### Bug Fixes

* let the range filter react to all user interaction like the other filters ([b844652](https://github.com/molgenis/molgenis-ui-filter/commit/b844652))

## [1.8.2](https://github.com/molgenis/molgenis-ui-filter/compare/v1.8.1...v1.8.2) (2020-04-30)


### Bug Fixes

* active filters not adjusted to array input ([1ac7c94](https://github.com/molgenis/molgenis-ui-filter/commit/1ac7c94))
* collapsed date picker for snapshot, added test and refactored cde ([2e4349e](https://github.com/molgenis/molgenis-ui-filter/commit/2e4349e))
* fixed values not being updated correctly when set in App.vue ([36a7034](https://github.com/molgenis/molgenis-ui-filter/commit/36a7034))
* probably some date thing breaks jenkins ([bf719dd](https://github.com/molgenis/molgenis-ui-filter/commit/bf719dd))
* removed snapshot for now, this is broken and needs manual intervention https://medium.com/front-end-weekly/jest-snapshot-testing-with-dates-and-times-f3badb8f1d87 ([0c24408](https://github.com/molgenis/molgenis-ui-filter/commit/0c24408))
* removed snapshot, because locale is not easiliy settable ([8d6a57d](https://github.com/molgenis/molgenis-ui-filter/commit/8d6a57d))
* restored snapshot ([a80779a](https://github.com/molgenis/molgenis-ui-filter/commit/a80779a))
* tests not taking locales into account ([001a981](https://github.com/molgenis/molgenis-ui-filter/commit/001a981))
* updated component, according to feedback. ([0814527](https://github.com/molgenis/molgenis-ui-filter/commit/0814527))

## [1.8.1](https://github.com/molgenis/molgenis-ui-filter/compare/v1.8.0...v1.8.1) (2020-04-21)


### Bug Fixes

* remove errors shown when a filter is still loading ([7827fe2](https://github.com/molgenis/molgenis-ui-filter/commit/7827fe2))

# [1.8.0](https://github.com/molgenis/molgenis-ui-filter/compare/v1.7.0...v1.8.0) (2020-04-21)


### Bug Fixes

* added unittest ([bd8f79e](https://github.com/molgenis/molgenis-ui-filter/commit/bd8f79e))


### Features

* Multifilter can now populate based on prefilled values. ([0c99ac5](https://github.com/molgenis/molgenis-ui-filter/commit/0c99ac5))

# [1.7.0](https://github.com/molgenis/molgenis-ui-filter/compare/v1.6.2...v1.7.0) (2020-04-21)


### Bug Fixes

* rangefilter will not force values on user ([6733c36](https://github.com/molgenis/molgenis-ui-filter/commit/6733c36))
* review changes ([a0e55ef](https://github.com/molgenis/molgenis-ui-filter/commit/a0e55ef))
* unit tests ([25f9fae](https://github.com/molgenis/molgenis-ui-filter/commit/25f9fae))


### Features

* clear range filter values and nice active filter ([ca0ac91](https://github.com/molgenis/molgenis-ui-filter/commit/ca0ac91))

## [1.6.2](https://github.com/molgenis/molgenis-ui-filter/compare/v1.6.1...v1.6.2) (2020-04-08)


### Bug Fixes

* Make only the cart title listen to drag events ([#48](https://github.com/molgenis/molgenis-ui-filter/issues/48)) ([b95ee66](https://github.com/molgenis/molgenis-ui-filter/commit/b95ee66))

## [1.6.1](https://github.com/molgenis/molgenis-ui-filter/compare/v1.6.0...v1.6.1) (2020-04-08)


### Bug Fixes


tests/mockdata.ts
/* istanbul ignore file */
// API setup for testing filters.
import axios from 'axios'
import { toRsqlValue } from '@molgenis/rsql'

const api = axios.create({
  baseURL: 'http://localhost:8080/api/',
  timeout: 15000
})

// Prevent triggering options multiple times.
let cache:any = { query: '', results: null }

const multifilterOptions = async (queryOptions : any) => {
  if (queryOptions === undefined) queryOptions = {}

  const nameAttribute = queryOptions.nameAttribute
  const queryType = queryOptions.queryType ? queryOptions.queryType : 'like'
  const query = queryOptions.query ? queryOptions.query : undefined
  const count = queryOptions.count ? queryOptions.count : 0

  let params:any = {}

  if (queryOptions && queryOptions.count > 0) {
    params.size = queryOptions.count
  }

  if (query) {
    params = { q: `${nameAttribute ? 'disease' : 'id'}=${queryType}=${queryType === 'in' ? `(${query})` : toRsqlValue(query)}` }
    if (query === cache.query) return cache.results
    cache.query = query
  }

  const data = await api.get(`/data/root_hospital_diagnosis`, { params })

  cache.results = data.data.items.map((i: any) => {
    return { value: i.data.id, text: i.data['disease'] }
  })

  return cache.results
}

const checkboxOptions =
  [
    { value: 'red', text: 'Red' },
    { value: 'green', text: 'Green' },
    { value: 'blue', text: 'Blue' }
  ]

const checkboxLotsOptions = [
  { value: 'red', text: 'Red' },
  { value: 'green', text: 'Green' },
  { value: 'blue', text: 'Blue' },
  { value: 'yellow', text: 'Yellow' },
  { value: 'white', text: 'White' },
  { value: 'purple', text: 'Purple' },
  { value: 'black', text: 'Black' }
]

const boolCheckboxes = [{ value: 'no', text: 'No' }, { value: 'yes', text: 'Yes' }]

export default { api, multifilterOptions, checkboxOptions, checkboxLotsOptions, boolCheckboxes }


tests/unit/setup.js
globalThis.IS_TEST = true


tests/unit/App.spec.ts
import { shallowMount, createLocalVue } from '@vue/test-utils'
import App from '@/App.vue'
import BootstrapVue from 'bootstrap-vue'
import store from '@/store'

const VueStash = require('vue-stash')

const localVue = createLocalVue()
localVue.use(VueStash.default)

describe('StringFilter.vue', () => {
  const wrapper = shallowMount(App, {
    localVue,
    parentComponent: { data: () => ({ store }) }
  })

  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})


.eslintrc.js
module.exports = {
  root: true,
  env: {
    jest: true,
    node: true
  },
  'extends': [
    'plugin:vue/recommended',
    '@vue/standard',
    '@vue/typescript'
  ],
  rules: {
    'curly': ['error', 'all'],
    'no-console': process.env.NODE_ENV === 'production' ? 'error' : 'off',
    'no-debugger': process.env.NODE_ENV === 'production' ? 'error' : 'off',
    '@typescript-eslint/member-delimiter-style': ['error', {
      multiline: {
        delimiter: 'none',
        requireLast: false
      },
      singleline: {
        delimiter: 'comma',
        requireLast: false
      }
    }],
    'vue/max-attributes-per-line': ['error', {
      'singleline': 3,
      'multiline': {
        'max': 1,
        'allowFirstLine': false
      }
    }],
    'vue/attributes-order': ['error', {
      'order': [
        'DEFINITION',
        'LIST_RENDERING',
        'CONDITIONALS',
        'RENDER_MODIFIERS',
        'GLOBAL',
        'UNIQUE',
        'TWO_WAY_BINDING',
        'OTHER_DIRECTIVES',
        'OTHER_ATTR',
        'EVENTS',
        'CONTENT'
      ],
      'alphabetical': true
    }]
  },
  parserOptions: {
    parser: '@typescript-eslint/parser',
    ecmaVersion: 2018,
    sourceType: 'module'
  },
  plugins: ['@typescript-eslint']
}


vue.config.js
const config = require('rc')('lifelines', {
  development: {
    proxy: 'https://master.dev.molgenis.org'
  }
})

module.exports = {
  configureWebpack: config => {
    if (process.env.NODE_ENV !== 'production') {
      config.devtool = process.env.DEVTOOL ? process.env.DEVTOOL : 'source-map'
      console.log(`Sourcemap mode: ${config.devtool}`)
    }
  },
  devServer: {
    // In CI mode, Safari cannot contact "localhost", so as a workaround, run the dev server using the jenkins agent pod dns instead.
    host: process.env.JENKINS_AGENT_NAME || 'localhost',
    proxy: process.env.NODE_ENV === 'production' ? undefined : {
      '^/api': {
        'target': config.development.proxy,
        'keepOrigin': true
      },
      '^/login': {
        'target': config.development.proxy,
        'changeOrigin': true
      }
    }
  }
}


jest.config.js
// Force use of modules in Babel when running tests
// with an external tool like vscode-jest.
process.env.VUE_CLI_BABEL_TARGET_NODE = true
process.env.VUE_CLI_BABEL_TRANSPILE_MODULES = true
process.env.BOOTSTRAP_VUE_NO_WARN = true
process.env.TZ = 'GMT'

module.exports = {
  moduleFileExtensions: [
    'js',
    'jsx',
    'json',
    'vue',
    'ts',
    'tsx'
  ],
  transform: {
    '^.+\\.vue$': 'vue-jest',
    '.+\\.(css|styl|less|sass|scss|svg|png|jpg|ttf|woff|woff2)$': 'jest-transform-stub',
    '^.+\\.tsx?$': 'ts-jest',
    '^.+\\.es.js$': 'babel-jest'
  },
  transformIgnorePatterns: [
    '/node_modules/(?!(@molgenis/molgenis-api-client.*))'
  ],
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1',
    '^.+.(css|styl|less|sass|scss|png|jpg|ttf|woff|woff2)$': 'jest-transform-stub'
  },
  snapshotSerializers: [
    'jest-serializer-vue'
  ],
  testMatch: [
    '**/tests/unit/**/*.spec.(js|jsx|ts|tsx)|**/__tests__/*.(js|jsx|ts|tsx)'
  ],
  testURL: 'http://localhost/',
  watchPlugins: [
    'jest-watch-typeahead/filename',
    'jest-watch-typeahead/testname'
  ],
  globals: {
    'ts-jest': {
      babelConfig: true
    }
  },
  collectCoverage: true
}


nightwatch.config.js
var packageJson = require('./package.json')

const buildName = packageJson.name + '#PR-' + process.env.CHANGE_ID + '-build-' + process.env.BUILD_NUMBER

module.exports = {

  test_settings: {
    chrome: {
      selenium: {
        cli_args: {
          'webdriver.chrome.driver': require('chromedriver').path
        }
      }
    },
    ci_chrome: {
      launch_url: 'http://ondemand.saucelabs.com:80',
      selenium_port: 80,
      selenium_host: 'ondemand.saucelabs.com',
      silent: true,
      username: process.env.SAUCE_CRED_USR,
      access_key: process.env.SAUCE_CRED_PSW,
      desiredCapabilities: {
        name: packageJson.name,
        build: buildName,
        'tunnel-identifier': process.env.TUNNEL_IDENTIFIER,
        browserName: 'chrome'
      }
    },
    ci_firefox: {
      launch_url: 'http://ondemand.saucelabs.com:80',
      selenium_port: 80,
      selenium_host: 'ondemand.saucelabs.com',
      silent: true,
      username: process.env.SAUCE_CRED_USR,
      access_key: process.env.SAUCE_CRED_PSW,
      desiredCapabilities: {
        name: packageJson.name,
        build: buildName,
        'tunnel-identifier': process.env.TUNNEL_IDENTIFIER,
        browserName: 'firefox'
      }
    },
    ci_ie11: {
      launch_url: 'http://ondemand.saucelabs.com:80',
      selenium_port: 80,
      selenium_host: 'ondemand.saucelabs.com',
      silent: true,
      username: process.env.SAUCE_CRED_USR,
      access_key: process.env.SAUCE_CRED_PSW,
      desiredCapabilities: {
        name: packageJson.name,
        build: buildName,
        'tunnel-identifier': process.env.TUNNEL_IDENTIFIER,
        browserName: 'internet explorer',
        platform: 'Windows 10',
        version: '11.103'
      }
    },
    ci_safari: {
      launch_url: 'http://ondemand.saucelabs.com:80',
      selenium_port: 80,
      selenium_host: 'ondemand.saucelabs.com',
      silent: true,
      username: process.env.SAUCE_CRED_USR,
      access_key: process.env.SAUCE_CRED_PSW,
      desiredCapabilities: {
        name: packageJson.name,
        build: buildName,
        'tunnel-identifier': process.env.TUNNEL_IDENTIFIER,
        browserName: 'safari'
      }
    },
    firefox: {
      desiredCapabilities: {
        browserName: 'firefox',
        javascriptEnabled: true,
        acceptSslCerts: true,
        marionette: true
      },
      selenium: {
        cli_args: {
          'webdriver.firefox.driver': require('geckodriver').path
        }
      }
    },
    safari: {
      desiredCapabilities: {
        browserName: 'safari',
        javascriptEnabled: true,
        acceptSslCerts: true
      }
    }
  }
}


src/store.ts
import axios from 'axios'
import data from '../tests/mockdata'

const api = axios.create({
  baseURL: 'http://localhost:8080/api/',
  timeout: 15000
})

// @ts-ignore
api.entity = 'root_hospital_diagnosis'

const providers = {
  multiFilter: async function (query:any) {
    // @ts-ignore
    const metadata = await api.get(`metadata/${api.entity}`)
    const nameAttr = metadata.data.data.attributes.items.filter((i:any) => i.data.labelAttribute).map((i) => i.data.name)[0]
    // @ts-ignore
    const data = await api.get(`data/${api.entity}?q=${nameAttr}=like=${query}`)
    return data.data.items.map((i:any) => {
      return { id: i.data.id, name: i.data[nameAttr] }
    })
  }
}

const state = {
  filters: {
    canEdit: true,
    available: [
      {
        active: true,
        label: 'DateTime filter',
        max: null,
        min: null,
        opens: 'right',
        range: true,
        selection: null, // range ? [null, null] : null
        time: true,
        type: 'DateTimeFilter'
      },
      {
        active: true,
        label: 'Multi filter',
        maxVisibleOptions: 5,
        options: [],
        provider: providers.multiFilter,
        search: '',
        selection: [],
        type: 'MultiFilter'
      },
      {
        active: false,
        description: 'search by string',
        label: 'Search',
        placeholder: 'placeholder',
        selection: 'test',
        type: 'StringFilter'
      },
      {
        active: true,
        bulkOperation: true,
        label: 'Color',
        maxVisibleOptions: 5,
        options: [],
        provider: () => {
          return new Promise((resolve) => {
            resolve(data.checkboxOptions)
          })
        },
        selection: [],
        type: 'CheckboxFilter'
      },
      {
        active: false,
        label: 'Number',
        max: 15,
        min: -5,
        selection: null,
        step: 1,
        type: 'NumberFilter',
        useSlider: false
      },
      {
        active: false,
        label: 'Age',
        max: 10,
        min: -10,
        selection: [null, null],
        step: 0.01,
        type: 'RangeFilter',
        useSlider: true
      },
      {
        active: false,
        description: 'Name of object',
        label: 'Name',
        selection: null,
        type: 'StringFilter'
      },
      {
        active: false,
        description: 'search by string',
        label: 'String',
        placeholder: 'placeholder',
        selection: null,
        type: 'StringFilter'
      },
      {
        active: false,
        bulkOperation: true,
        label: 'Checkbox',
        maxVisibleOptions: 1,
        options: [],
        provider: () => {
          return new Promise((resolve) => {
            resolve(data.checkboxOptions)
          })
        },
        selection: [],
        type: 'CheckboxFilter'
      },
      {
        active: false,
        bulkOperation: true,
        label: 'Way too long name to really fit in the user interface',
        maxVisibleOptions: 1,
        options: [],
        provider: () => {
          return new Promise((resolve) => {
            resolve(data.checkboxOptions)
          })
        },
        selection: null,
        type: 'CheckboxFilter'
      }
    ]
  }
}

let id = 1
for (const filter of state.filters.available) {
  Object.assign(filter, {
    id,
    collapsed: true,
    collapsible: true
  })

  id += 1
}

export default state
